command/ssh/certificate.go

```diff
@@ -0,0 +1,391 @@
+			return err
+		}
+	}
+
+	caClient, err := flow.GetClient(ctx, subject, token)
+	if err != nil {
+		return err
+	}
+
+	var sshPub ssh.PublicKey
+	var pub, priv interface{}
+
+	if isSign {
+		// Used given public key
+		in, err := utils.ReadFile(keyFile)
+		if err != nil {
+			return err
+		}
+
+		sshPub, _, _, _, err = ssh.ParseAuthorizedKey(in)
+		if err != nil {
+			return errors.Wrap(err, "error parsing public key")
+		}
+	} else {
+		// Generate keypair
+		pub, priv, err = keys.GenerateDefaultKeyPair()
+		if err != nil {
+			return err
+		}
+
+		sshPub, err = ssh.NewPublicKey(pub)
+		if err != nil {
+			return errors.Wrap(err, "error creating public key")
+		}
+	}
+
+	var sshAuPub ssh.PublicKey
+	var sshAuPubBytes []byte
+	var auPub, auPriv interface{}
+	if isAddUser {
+		auPub, auPriv, err = keys.GenerateDefaultKeyPair()
+		if err != nil {
+			return err
+		}
+		sshAuPub, err = ssh.NewPublicKey(auPub)
+		if err != nil {
+			return errors.Wrap(err, "error creating public key")
+		}
+		sshAuPubBytes = sshAuPub.Marshal()
+	}
+
+	resp, err := caClient.SignSSH(&api.SignSSHRequest{
+		PublicKey:        sshPub.Marshal(),
+		OTT:              token,
+		Principals:       principals,
+		CertType:         certType,
+		ValidAfter:       validAfter,
+		ValidBefore:      validBefore,
+		AddUserPublicKey: sshAuPubBytes,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Write files
+	if !isSign {
+		// Private key (with password unless --no-password --insecure)
+		opts := []pemutil.Options{
+			pemutil.ToFile(keyFile, 0600),
+		}
+		switch {
+		case noPassword && insecure:
+		case passwordFile != "":
+			opts = append(opts, pemutil.WithPasswordFile(passwordFile))
+		default:
+			opts = append(opts, pemutil.WithPasswordPrompt("Please enter the password to encrypt the private key"))
+		}
+		_, err = pemutil.Serialize(priv, opts...)
+		if err != nil {
+			return err
+		}
+
+		if err := utils.WriteFile(pubFile, marshalPublicKey(sshPub, subject), 0644); err != nil {
+			return err
+		}
+	}
+
+	// Write certificate
+	if err := utils.WriteFile(crtFile, marshalPublicKey(resp.Certificate, subject), 0644); err != nil {
+		return err
+	}
+
+	// Write Add User keys and certs
+	if isAddUser {
+		id := provisioner.SanitizeSSHUserPrincipal(subject) + "-provisioner"
+		if _, err := pemutil.Serialize(auPriv, pemutil.ToFile(baseName+"-provisioner", 0600)); err != nil {
+			return err
+		}
+		if err := utils.WriteFile(baseName+"-provisioner.pub", marshalPublicKey(sshAuPub, id), 0644); err != nil {
+			return err
+		}
+		if err := utils.WriteFile(baseName+"-provisioner-cert.pub", marshalPublicKey(resp.AddUserCertificate, id), 0644); err != nil {
+			return err
+		}
+	}
+
+	if !isSign {
+		ui.PrintSelected("Private Key", keyFile)
+		ui.PrintSelected("Public Key", pubFile)
+	}
+	ui.PrintSelected("Certificate", crtFile)
+
+	// Attempt to add key to agent
+	if err := sshAddKeyToAgent(subject, resp.Certificate.Certificate, priv); err != nil {
```

Comment:
Isn't it possible that `priv` is `nil` here? If `isSign` then we never set `priv`.

Reply:
The answer to that question is 'yes'. As a fix I've added another flag (--private-key) so that you can still specify the private key for adding to ssh agent. If `priv == nil` then we won't attempt to add to the ssh agent.